torch/csrc/Stream.cpp

```diff
@@ -256,6 +261,112 @@ static PyObject* THPStream_eq(THPStream* self, THPStream* other) {
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* THPStream_enter(PyObject* _self, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  auto self = (THPStream*)_self;
+  c10::DeviceType stream_device_type =
+      static_cast<c10::DeviceType>(self->device_type);
+  // No operation is performed if the stream does not belong to an accelerator.
+  if (C10_UNLIKELY(!at::accelerator::isAccelerator(stream_device_type))) {
+    // Support with torch.Stream() as stream: works
+    Py_INCREF(_self);
+    return _self;
+  }
+  c10::DeviceIndex cur_device_idx = at::accelerator::getDeviceIndex();
+  c10::DeviceIndex stream_device_idx =
+      static_cast<c10::DeviceIndex>(self->device_index);
+  // If the stream is not on the current device, switch the current device to
+  // the device of the stream.
+  if (stream_device_idx != cur_device_idx) {
+    at::accelerator::setDeviceIndex(stream_device_idx);
+  }
+  c10::Stream cur_stream = at::accelerator::getCurrentStream(stream_device_idx);
+  at::accelerator::setCurrentStream(c10::Stream::unpack3(
+      self->stream_id, stream_device_idx, stream_device_type));
+  // Save the current device index and previous stream to the context.
+  auto ctx_device_index = py::reinterpret_steal<py::object>(
+      THPUtils_packDeviceIndex(cur_device_idx));
+  auto ctx_stream =
+      py::reinterpret_steal<py::object>(THPStream_Wrap(cur_stream));
+  PyObject_SetAttrString(_self, "_ctx_device_index", ctx_device_index.ptr());
```

Comment:
Please delete these fields on exit and check here that they're not already set.
We do want to raise an error for example if the user passes the stream object to two threads and try to enter it from both at the same time.
We can look into fixing this behavior as well if that's something we care about. But a nice error is ok for now.

Reply:
I add a `TORCH_CHECK` to ensure `context` is `nullptr` when it start to `__enter__`.